src/main/java/org/cbioportal/service/impl/ClinicalDataServiceImpl.java

```diff
@@ -227,17 +234,44 @@ public List<ClinicalData> getPatientClinicalDataDetailedToSample(List<String> st
     }
 
     @Override
-    public List<ClinicalData> fetchSampleClinicalTable(List<String> studyIds, List<String> sampleIds, Integer pageSize,
-                                                       Integer pageNumber, String searchTerm, String sortBy,
-                                                       String direction) {
-        return clinicalDataRepository.fetchSampleClinicalTable(studyIds, sampleIds, pageSize, pageNumber, searchTerm,
-                                                                            sortBy, direction);
-    }
+    public ImmutablePair<SampleClinicalDataCollection, Integer> fetchSampleClinicalTable(List<String> studyIds, List<String> sampleIds, Integer pageSize, Integer pageNumber, String searchTerm, String sortBy, String direction) {
 
-    @Override
-    public Integer fetchSampleClinicalTableCount(List<String> studyIds, List<String> sampleIds,
-                                                 String searchTerm, String sortBy, String direction) {
-        return clinicalDataRepository.fetchSampleClinicalTableCount(studyIds, sampleIds, searchTerm, 
-            sortBy, direction);
+        SampleClinicalDataCollection sampleClinicalDataCollection = new SampleClinicalDataCollection();
+        if (studyIds == null || studyIds.isEmpty() || sampleIds == null || sampleIds.isEmpty()) {
+            return new ImmutablePair<>(sampleClinicalDataCollection, 0);
+        }
+        
+        // Request un-paginated data.
+        List<Integer> allSampleInternalIds = clinicalDataRepository.getVisibleSampleInternalIdsForClinicalTable(
+            studyIds, sampleIds,
+            null, null,
+            searchTerm, sortBy, direction
+        );
+        Integer offset = paginationCalculator.offset(pageSize, pageNumber);
+
+        if (allSampleInternalIds.isEmpty() || offset >= allSampleInternalIds.size()) {
+            return new ImmutablePair<>(sampleClinicalDataCollection, 0);
+        }
+
+        // Apply pagination to the sampleId list.
+        Integer toIndex = paginationCalculator.lastIndex(offset, pageSize, allSampleInternalIds.size());
+        List<Integer> visibleSampleInternalIds = allSampleInternalIds.subList(offset, toIndex);
+
+        List<ClinicalData> sampleClinicalData = clinicalDataRepository.getSampleClinicalDataBySampleInternalIds(
+            visibleSampleInternalIds
+        );
+        List<ClinicalData> patientClinicalData = clinicalDataRepository.getPatientClinicalDataBySampleInternalIds(
+            visibleSampleInternalIds
+        );
+        
+        // Merge sample and patient level clinical data and key by unique sample-key.
+        sampleClinicalDataCollection.setByUniqueSampleKey(
+            Stream.concat(sampleClinicalData.stream(), patientClinicalData.stream())
+                .collect(Collectors.groupingBy(clinicalDatum ->
+                    calculateBase64(clinicalDatum.getSampleId(), clinicalDatum.getStudyId())
+        )));
```

Comment:
Just from observation. Our functions are usually bigger than they need to be. We should start focusing on keeping them small. With that, I would extract this out into a fn maybe buildSampleClinicalDataCollection